File: 2020/day09.py
from os.path import isfile, join as path_join
from typing import Callable

file_name = path_join('input', 'day09.txt')


def to_list(mf: Callable = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]


def to_gen(mf: Callable = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)


if not isfile(file_name):
	from aoc import get_input_file

	get_input_file(session_path=['..', '.env'])

data = to_list()


def part1():
	pre = data[:25]

	pre_set = set(data[:25])

	for i in range(25, len(data)):
		num = data[i]

		valid = False

		for v in pre_set:
			res = num - v
			if res != v and res in pre_set:
				valid = True
				break

		if not valid:
			return num

		pre_set.remove(pre[0])
		pre_set.add(num)

		pre.pop(0)
		pre.append(num)


def part2(invalid):
	max_len = 0
	max_i = 0

	i = 0
	j = 0
	s = data[i]

	while i < len(data):
		if s < invalid:
			j += 1
			if j < len(data):
				s += data[j]

		elif s > invalid:
			s -= data[i]
			i += 1

		else:
			if j - i + 1 > max_len:
				max_len = j - i + 1
				max_i = i

			s -= data[i]
			i += 1

	print(min(data[max_i:max_i + max_len]) + max(data[max_i:max_i + max_len]))


invalid = part1()
print(invalid)

part2(invalid)

File: cpp/2020/day09.cpp
#include "../aocHelper.h"

class Day09 : public BaseDay
{
public:
	Day09() : BaseDay("09") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		long long n;

		vector<long long> data;

		deque<long long> pre;

		unordered_set<long long> pre_set;

		for (int i = 0; i < 25; i++)
		{
			int n = numericParse<long long>(input);

			data.push_back(n);
			pre.push_back(n);
			pre_set.insert(n);
		}

		while ((n = numericParse<long long>(input)))
		{
			data.push_back(n);

			bool valid = false;

			for (auto& v : pre_set)
			{
				long long res = n - v;
				if (res != v && pre_set.count(res))
				{
					valid = true;
					break;
				}
			}

			if (!valid)
			{
				part1 = n;
				goto section2;
			}

			pre_set.erase(pre.front());
			pre_set.insert(n);

			pre.pop_front();
			pre.push_back(n);
		}

	section2:

		int max_len = 0, max_i = 0;
		int i = 0, j = 0;
		long long s = data.front();

		while (i < data.size())
		{
			if (s < part1)
			{
				j++;
				if (j < data.size())
				{
					s += data[j];
				}
			}
			else if (s > part1)
			{
				s -= data[i];
				i++;
			}
			else
			{
				if (j - i + 1 > max_len)
				{
					max_len = j - i + 1;
					max_i = i;
				}

				s -= data[i];
				i++;
			}
		}

		long long min = *min_element(data.begin() + max_i, data.begin() + max_i + max_len);
		long long max = *max_element(data.begin() + max_i, data.begin() + max_i + max_len);

		part2 = min + max;

		return { part1, part2 };
	}
};
